lib/core/Axios.js

```diff
@@ -45,20 +47,62 @@ Axios.prototype.request = function request(config) {
     config.method = 'get';
   }
 
-  // Hook up interceptors middleware
-  var chain = [dispatchRequest, undefined];
-  var promise = Promise.resolve(config);
+  var requestCancelled = config.cancelToken && config.cancelToken.reason;
 
+  // filter out skipped interceptors
+  var requestInterceptorChain = [];
+  var synchronousChain = [];
   this.interceptors.request.forEach(function unshiftRequestInterceptors(interceptor) {
-    chain.unshift(interceptor.fulfilled, interceptor.rejected);
+    if (typeof interceptor.runWhen === 'function' && interceptor.runWhen(config) === false) {
+      return;
+    }
+
+    synchronousChain.push(interceptor.synchronous);
+
+    requestInterceptorChain.push(interceptor.fulfilled, interceptor.rejected);
   });
 
+  var synchronousRequestInterceptors = synchronousChain.every(function(interceptor) {
+    return interceptor === true;
+  });
+
+  var responseInterceptorChain = [];
   this.interceptors.response.forEach(function pushResponseInterceptors(interceptor) {
-    chain.push(interceptor.fulfilled, interceptor.rejected);
+    responseInterceptorChain.push(interceptor.fulfilled, interceptor.rejected);
   });
 
-  while (chain.length) {
-    promise = promise.then(chain.shift(), chain.shift());
+  var promise;
+
+  if (customAdapter || requestCancelled || !synchronousRequestInterceptors) {
```

Comment:
I'm using the async chain for the `customAdapter` because the custom adapter function provided by the consumer is not guaranteed to return a Promise. That adapter is called with a `.then` here: https://github.com/axios/axios/blob/master/lib/core/dispatchRequest.js#L52
By not wrapping it with a promise would cause a `TypeError` to be thrown on that line.

Reply:
For the cancelled request -> this is in order for the error to be caught in the catch block in the event that the request is cancelled before it is executed. This is the unit tests that captures this scenario: https://github.com/axios/axios/blob/master/test/specs/cancel.spec.js#L13

Reply:
According to https://github.com/axios/axios/blob/master/lib/adapters/README.md, `customAdapter` should also return a Promise. Or why does the old implementation work well?

Reply:
You are correct, the docs specify that a Promise should be returned from the `customAdapter` but it's not currently enforced in the code. Even the unit test for the `customAdapter` breaks that rule: https://github.com/axios/axios/blob/master/test/specs/adapter.spec.js#L9

It was still working in the old implementation because of the pre-emptive promise creation on every Axios request. I wanted to avoid breaking any existing user functionality, that's why I'm checking for the presence of a `customAdapter` and creating a promise.

Reply:
That test case is not a valid use case. In fact, an error will be thrown. "TypeError: Cannot read property 'then' of undefined". So I think it has been enforced in the code.

```
const axios = require('axios');

axios({
    url: 'sb.com',
    adapter() {
        console.log('here')
    }
}).catch(console.log)

// output
here
TypeError: Cannot read property 'then' of undefined
    at dispatchRequest (/Users/zhong/Desktop/test/axios/node_modules/axios/lib/core/dispatchRequest.js:52:25)
```

Reply:
Interesting! I'm running this code on master and the test is passing:

```
var axios = require('../../index');

describe('adapter', function () {
  it('should support custom adapter', function (done) {
    axios({
      url: 'sb.com',
      adapter: function() {
        console.log('here');
      }
    }).catch(console.log);

    setTimeout(function () {
      done();
    }, 100);
  });
});
```

It does go into the catch block however and logs the error: 

![image](https://user-images.githubusercontent.com/22783686/73848454-2797ad00-47f6-11ea-847e-f8e414a75c41.png)

Reply:
But you are right, by going into the catch block -> it makes it an invalid use case that consumers should not rely on. I will modify the unit test accordingly.

Reply:
You have added the catch wrapper for `dispatchRequest`, shall we also check the cancelled request here?

Reply:
Hmmm... I'm actually not sure if that will ever catch anything. Because if the request is canceled before being sent -> it happens in the sync flow. And if it's canceled afterward, it happens here: 

![image](https://user-images.githubusercontent.com/22783686/74040351-8d6c6c00-4991-11ea-85fd-08660fe1b779.png)

Reply:
Yes, so maybe we can remove the check.

Reply:
Check has been removed.

Reply:
Wait a minute. What I want to remove is `if (requestCancelled`. The catch wrapper of dispatchRequest is still useful as inside it, maybe somewhere like transformRequest can also throw an error.

Reply:
Oh, my bad. I misunderstood! We actually cannot remove the `if (requestCancelled` check. This check exists in order for the error to be caught in the catch block in the event that the request is canceled before it is executed. This is the unit test that captures this scenario: https://github.com/axios/axios/blob/master/test/specs/cancel.spec.js#L13

Reply:
The unit test is failed because you didn't wrap `dispatchRequest` by the catch. Now you have it, so `requestCancelled` can be removed now. I pulled your codes and confirmed that.

Reply:
Ah, you are right. I will remove the check.